src/Sarif/Visitors/RebaseUriVisitor.cs

```diff
@@ -0,0 +1,187 @@
+                    baseUriDictionary = new Dictionary<string, Uri>();
+                }
+            }
+            
+            // Note--this is an add or update, so if this is run twice with the same base variable, we'll replace the path.
+            baseUriDictionary[_baseName] = _baseUri;
+            
+            newRun.Properties[BaseUriDictionaryName] = ReserializePropertyDictionary(baseUriDictionary);
+
+            return newRun;
+        }
+
+        /// <summary>
+        /// If we are changing the URIs in Results to be relative, we need to also change the URI keys in the files dictionary
+        /// to be relative.
+        /// 
+        /// For FileData, we need to fix up the URI data (making it relative to the appropriate base address), 
+        /// and also fix up the ParentKey (as we are patching the keys in the files dictionary in the Run).
+        /// (We need to fix up the keys as we are patching the PhysicalLocation in the Result objects.)
+        /// </summary>
+        /// <param name="run">A run to fix the Files dictionary of.</param>
+        internal void FixFiles(Run run)
+        {
+            Dictionary<string, FileData> newDictionary = new Dictionary<string, FileData>();
+
+            foreach (var key in run.Files.Keys)
+            {
+                Uri oldUri;
+                string newKey = key;
+                FileData data = run.Files[key];
+                // If the old uri is absolute and we need to rebase it, we should.
+                if (Uri.TryCreate(key, UriKind.Absolute, out oldUri) && oldUri.IsAbsoluteUri && _baseUri.IsBaseOf(oldUri))
+                {
+                    Uri newUri = _baseUri.MakeRelativeUri(oldUri);
+
+                    // Ensure the filedata reflects the correct base URI details.
+                    if (data != null)
+                    {
+                        if (data.ParentKey != null)
+                        {
+                            Uri parentUri;
+                            if (Uri.TryCreate(data.ParentKey, UriKind.Absolute, out parentUri))
+                            {
+                                if (oldUri.IsAbsoluteUri && _baseUri.IsBaseOf(parentUri))
+                                {
+                                    data.ParentKey = _baseUri.MakeRelativeUri(new Uri(data.ParentKey)).ToString();
+                                }
+                            }
+                        }
+
+                        data.Uri = newUri;
+                        data.UriBaseId = _baseName;
+                    }
+
+                    if(newDictionary.ContainsKey(newUri.ToString()))
+                    {
+                        throw new InvalidOperationException("Cannot rebase this file, as two URIs will collide in the file dictionary.");
+                    }
+                    newKey = newUri.ToString();
```

Comment:
>newKey = newUri.ToString() [](start = 20, length = 26)

this code makes me realize we could attempt to normalize URL keys to avoid issues where URLs differ in rendering but aren't distinct when properly encoded or otherwise normalized. maybe this is a new action?